src/util/amount.rs

```diff
@@ -0,0 +1,367 @@
+    }
+}
+
+impl<P: Precision> ops::Sub for Amount<P> {
+    type Output = Amount<P>;
+
+    fn sub(self, rhs: Amount<P>) -> Self::Output {
+        Amount::from_inner(self.0 - rhs.0)
+    }
+}
+
+impl<P: Precision> ops::Mul<i64> for Amount<P> {
+    type Output = Amount<P>;
+
+    fn mul(self, rhs: i64) -> Self::Output {
+        Amount::from_inner(self.0 * rhs)
+    }
+}
+
+impl<P: Precision> ops::Div for Amount<P> {
+    type Output = f64;
+
+    fn div(self, rhs: Amount<P>) -> Self::Output {
+        self.0 as f64 / rhs.0 as f64
+    }
+}
+
+impl<P: Precision> ops::Div<f64> for Amount<P> {
+    type Output = Amount<P>;
+
+    fn div(self, rhs: f64) -> Self::Output {
+        Amount::from_inner((self.0 as f64 / rhs as f64) as i64)
+    }
+}
+
+impl PartialEq for Amount {
+    fn eq(&self, other: &Amount) -> bool {
+        PartialEq::eq(&self.0, &other.0)
+    }
+}
+impl Eq for Amount {}
+
+impl PartialOrd for Amount {
+    fn partial_cmp(&self, other: &Amount) -> Option<::std::cmp::Ordering> {
+        PartialOrd::partial_cmp(&self.0, &other.0)
+    }
+}
+
+impl Ord for Amount {
+    fn cmp(&self, other: &Amount) -> ::std::cmp::Ordering {
+        Ord::cmp(&self.0, &other.0)
+    }
+}
+
+/// An error during `Amount` parsing.
+#[derive(Debug)]
+pub struct ParseAmountError(std::num::ParseFloatError);
+
+impl fmt::Display for ParseAmountError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "parse amount error: {}", self.0)
+    }
+}
+
+impl error::Error for ParseAmountError {
+    fn cause(&self) -> Option<&error::Error> {
+        Some(&self.0)
+    }
+
+    fn description(&self) -> &'static str {
+        "parse amount error"
+    }
+}
+
+impl FromStr for Amount {
+    type Err = ParseAmountError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let btc = f64::from_str(s).map_err(ParseAmountError)?;
+        Ok(btc.into_btc())
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Amount {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        Ok(Amount::from_inner(Inner::deserialize(deserializer)?))
+    }
+}
+
+#[cfg(feature = "serde")]
+impl serde::Serialize for Amount {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        Inner::serialize(&self.0, serializer)
+    }
+}
+
+//TODO(stevenroose) this trait in itself is currently of limited use since the first generic
+//parameter is always unknown, so always the awkward notation is to be used
+pub trait IntoAmount<F: Precision, T: Precision> {
+    fn into_amount(self) -> Amount<T>;
+}
+
+impl<F: Precision, T: Precision> IntoAmount<F, T> for f64 {
+    fn into_amount(self) -> Amount<T> {
+        Amount::from_inner(pow_round_and_to_inner(self, T::scale() - F::scale()))
+    }
+}
+
+impl<'a, F: Precision, T: Precision> IntoAmount<F, T> for &'a f64 {
```

Comment:
Why is this also implemented for `&f64` and not only for `f64`? Since `f64` is `Copy` that shouldn't be much of a problem. I think auto deref should take care of making the appropriate copies.

Reply:
I just copied this from rust-amount. I figured if it existed there, might have been a reason for it..